docs(solid-testing-library): add solid testing library docs (#1370)
testing-library/testing-library-docs: docs/solid-testing-library/api.mdx

---
id: api
title: API
sidebar_label: API
---

Due to being inspired by the [preact-testing-library][preact-docs] you can check
its page for more information.

[preact-docs]: ../preact-testing-library/intro.mdx

There are several key differences, to be aware of.

> ## Differences from the Preact Testing Library

Reviewer: I don't know if this is needed.
Most Testing Libraries are all inspired by DOM testing library/each other.

Thoughts?

Reviewer 2: I think that this doc is mainly showing comparisons (as can be seen in the `render` section, I'm still not sure it's needed though, I believe we can remove the "preact" examples and just keep the "solid" examples here.. Wdyt?

Author: done

> ## Differences from the Preact Testing Library

Reviewer 2: We usually also keep a table of contents for all the API's in this file with links (these can be auto generated AFAIR) :)

Author: adding a Toc

Author: Thank you :)

### `render`

The `render` function takes in a function that returns a Solid Component, rather
than simply the component itself.

```tsx
// With @testing-library/preact
const results = render(<YourComponent />, options)
```

```tsx
// With solid-testing-library
const results = render(() => <YourComponent />, options)
```

Solid.js does _not_ re-render, it merely executes side effects triggered by
reactive state that change the DOM, therefore there is no `rerender` method. You
can use global signals to manipulate your test component in a way that causes it
to update.

> In extension of the original API, the render function of this testing library

Reviewer 2: This is really a nit:

```suggestion
In addition to the original API, the render function of this testing library
```

supports a convenient `location` option that will set up an in-memory router

> pointing at the specified location. Since this setup is not instantaneous, you

Reviewer 2: Is this the correct meaning? :)
```suggestion
pointing at the specified location. Since this setup is not synchronous, you
```

need to first use asynchronous queries (`findBy`) after employing it:

```tsx
it('uses params', async () => {
  const App = () => (
    <Routes>
      <Route path="/ids/:id" component={() => <p>Id: {useParams()?.id}</p>} />
      <Route path="/" component={() => <p>Start</p>} />
    </Routes>
  )
  const {findByText} = render(() => <App />, {location: 'ids/1234'})
  expect(await findByText('Id: 1234')).not.toBeFalsy()
})
```

It uses `@solidjs/router`, so if you want to use a different router, you should

> consider the `wrapper` option instead. If you attempt to use this without having

Reviewer 2: I think that this information should also be added to an FAQ but this is just an opinion and not a blocking comment :)

Author: I would say this information is relevant to keep on this part as it is something that impacts the usage of this render function. 🤔

Reviewer 2: I agree on this one, that's why I said also, but we can iterate on this later if we'll find the need :)
Thanks!

the package installed, you will receive an error message.

### `renderHook`

Solid.js external reactive state does not require any DOM elements to run in, so
our `renderHook` call to test hooks in the context of a component (if your hook
does not require the context of a component, `createRoot` should suffice to test

> the reactive behavior; for convenience, we also have `createEffect`, which is

Reviewer 2: Maybe we should link to `createEffect` here? and then we'll be able to remove the "which is described later" sentence.

Author: will do :)

return value. Instead, it has an `owner` to be used with
[`runWithOwner`](https://www.solidjs.com/docs/latest/api#runwithowner) if
required. It also exposes a `cleanup` function, though this is already
automatically called after the test is finished.

```ts
function renderHook<Args extends any[], Result>(
  hook: (...args: Args) => Result,
  options: {
    initialProps?: Args,
    wrapper?: Component<{ children: JSX.Element }>
  }
) => {
  result: Result;
  owner: Owner | null;
  cleanup: () => void;
}
```

This can be used to easily test a hook / primitive:

```ts
const {result} = renderHook(createResult)
expect(result).toBe(true)
```

If you are using a `wrapper` with `renderHook`, make sure it will **always**
return `props.children` - especially if you are using a context with
asynchronous code together with `<Show>`, because this is required to get the
value from the hook and it is only obtained synchronously once and you will
otherwise only get `undefined` and wonder why this is the case.

Solid.js supports
[custom directives](https://www.solidjs.com/docs/latest/api#use___), which is a
convenient pattern to tie custom behavior to elements, so we also have a

> `renderDirective` call, which augments `renderHook` to take a directive as first

Reviewer 2: Should this also have an `h3`?

> Solid.js supports
> [custom directives](https://www.solidjs.com/docs/latest/api#use___), which is a
> convenient pattern to tie custom behavior to elements, so we also have a
> `renderDirective` call, which augments `renderHook` to take a directive as first
> argument, accept an `initialValue` for the argument and a `targetElement`
> (string, HTMLElement or function returning a HTMLElement) in the `options` and
> also returns `arg` and `setArg` to read and manipulate the argument of the
> directive.

Reviewer: ```suggestion
Solid.js supports
[custom directives](https://www.solidjs.com/docs/latest/api#use___), which is a
convenient pattern to tie custom behavior to elements, so we also have a
`renderDirective` call, which augments `renderHook` to take a directive as first
argument, accept an `initialValue` for the argument and a `targetElement`
(string, HTMLElement or function returning an HTMLElement) in the `options` and
also returns `arg` and `setArg` to read and manipulate the argument of the
directive.
```
